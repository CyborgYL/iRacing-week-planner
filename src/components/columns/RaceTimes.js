// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import type { TimeableRace } from '../../lib/races';
import Modal from '../modal/Modal';
import styles from './columns.scss';

import bootstrapStyles from '../../styles/main.module.scss';

type Props = {
  race: {
    ...TimeableRace,
    series: string,
  },
};

export default function NextRace({ race }: Props) {
  const [modalOpen, setModalOpen] = React.useState(false);
  const openModal = () => setModalOpen(true);
  const closeModal = () => setModalOpen(false);
  const { t } = useTranslation();

  if (race.setTimes) {
    const weekStart = moment(moment().utc()).subtract(1, 'days').startOf('isoWeek').add(1, 'days');

    return (
      <td className={styles.clickableCell}>
        <button type="button" className={styles.cellButton} onClick={openModal} onKeyPress={openModal}>
          {t('Set Times')}
        </button>
        <Modal
          isOpen={modalOpen}
          onClose={closeModal}
          title={t('Set times for {{series}}', { series: t(race.series) })}
          doneAction={closeModal}
        >
          <div className={bootstrapStyles['container-fluid']}>
            <ul>
              {race.setTimes.map(
                (time) => (
                  <li key={time}>
                    {t('{{timeLocal, ddd h:mma}} ({{timeUtc, ddd h:mma z}})', {
                      timeLocal: moment(weekStart).add(time).local().toDate(),
                      timeUtc: moment(weekStart).add(time).utc().toDate(),
                    })}
                  </li>
                ),
              )}
            </ul>
          </div>
        </Modal>
      </td>
    );
  }

  if (race.everyTime) {
    return (
      <td>
        <div>
          {t('Every {{every}} starting at {{time, H:mm}} UTC', {
            every: race.everyTime.humanize().replace(/an?\s/, ''),
            time: moment().utc().startOf('day').add(race.offset)
              .toDate(),
          })}
        </div>
      </td>
    );
  }

  return <td>{t('No time data')}</td>;
}
