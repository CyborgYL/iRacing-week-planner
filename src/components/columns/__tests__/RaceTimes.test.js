import moment from 'moment';
import React from 'react';
import { shallow } from 'enzyme';
import { describe, test, expect } from '@jest/globals';
import Modal from '../../modal/Modal';

import RaceTimes from '../RaceTimes';

describe('components/columns/RaceTimes', () => {
  test('renders no time data', () => {
    const component = shallow(<RaceTimes
      race={{
        raceTimes: null,
      }}
    />);

    expect(component).toMatchSnapshot();
  });

  test('renders settimes', () => {
    const component = shallow(<RaceTimes
      race={{
        series: 'Mazda Cup',
        setTimes: [
          moment.duration({ days: 3, hours: 2 }),
          moment.duration({ days: 4, hours: 13 }),
        ],
      }}
    />);

    expect(component).toMatchSnapshot();

    component.first().simulate('click');

    expect(component).toMatchSnapshot();

    component.find(Modal).first().props().onClose();

    expect(component).toMatchSnapshot();
  });

  test('renders everytime', () => {
    const component = shallow(<RaceTimes
      race={{
        series: 'Mazda Cup',
        everyTime: moment.duration({ hours: 3 }),
        offset: moment.duration({ minutes: 90 }),
      }}
    />);

    expect(component).toMatchSnapshot();
  });
});
