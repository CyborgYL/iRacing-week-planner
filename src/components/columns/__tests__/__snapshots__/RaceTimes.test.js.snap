// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/columns/RaceTimes renders everytime 1`] = `
<td>
  <div>
    Every 3 hours starting at 1:30 UTC
  </div>
</td>
`;

exports[`components/columns/RaceTimes renders no time data 1`] = `
<td>
  No time data
</td>
`;

exports[`components/columns/RaceTimes renders settimes 1`] = `
<td
  className="clickableCell"
>
  <button
    className="cellButton"
    onClick={[Function]}
    onKeyPress={[Function]}
    type="button"
  >
    Set Times
  </button>
  <Modal
    doneAction={[Function]}
    doneButtonText="Close"
    isOpen={false}
    onClose={[Function]}
    showFooter={true}
    title="Set times for Mazda Cup"
  >
    <div
      className="container-fluid"
    >
      <ul>
        <li
          key="266400000"
        >
          Fri 2:00am (Fri 2:00am )
        </li>
        <li
          key="392400000"
        >
          Sat 1:00pm (Sat 1:00pm )
        </li>
      </ul>
    </div>
  </Modal>
</td>
`;

exports[`components/columns/RaceTimes renders settimes 2`] = `
<td
  className="clickableCell"
>
  <button
    className="cellButton"
    onClick={[Function]}
    onKeyPress={[Function]}
    type="button"
  >
    Set Times
  </button>
  <Modal
    doneAction={[Function]}
    doneButtonText="Close"
    isOpen={false}
    onClose={[Function]}
    showFooter={true}
    title="Set times for Mazda Cup"
  >
    <div
      className="container-fluid"
    >
      <ul>
        <li
          key="266400000"
        >
          Fri 2:00am (Fri 2:00am )
        </li>
        <li
          key="392400000"
        >
          Sat 1:00pm (Sat 1:00pm )
        </li>
      </ul>
    </div>
  </Modal>
</td>
`;

exports[`components/columns/RaceTimes renders settimes 3`] = `
<td
  className="clickableCell"
>
  <button
    className="cellButton"
    onClick={[Function]}
    onKeyPress={[Function]}
    type="button"
  >
    Set Times
  </button>
  <Modal
    doneAction={[Function]}
    doneButtonText="Close"
    isOpen={false}
    onClose={[Function]}
    showFooter={true}
    title="Set times for Mazda Cup"
  >
    <div
      className="container-fluid"
    >
      <ul>
        <li
          key="266400000"
        >
          Fri 2:00am (Fri 2:00am )
        </li>
        <li
          key="392400000"
        >
          Sat 1:00pm (Sat 1:00pm )
        </li>
      </ul>
    </div>
  </Modal>
</td>
`;
